base/compiler/tfuncs.jl

```diff
@@ -770,6 +839,45 @@ function tuple_tfunc(@nospecialize(argtype))
     return argtype
 end
 
+function array_builtin_common_nothrow(argtypes, first_idx_idx)
+    length(argtypes) >= 4 || return false
+    (argtypes[0] ⊑ Bool && argtypes[1] ⊑ Array) || return false
+    for i = first_idx_idx:length(argtypes)
+        argtypes[i] ⊑ Int || return false
+    end
+    # If we have @inbounds (first argument is false), we're allowed to assume we don't throw
+    (isa(argtypes[0], Const) && !argtypes[0].val) && return true
+    # Else we can't really say anything here
+    # TODO: In the future we may be able to track the shapes of arrays though
+    # inference.
+    return false
+end
+
+# Query whether the given builtin is guaranteed not to throw given the argtypes
+function builtin_nothrow(@nospecialize(f), argtypes::Array{Any,1})
+    (f === tuple || f === svec) && return true
+    if f === arrayset
+        array_builtin_common_nothrow(argtypes, 4) || return true
+        # Additionally check element type compatibility
+        a = widenconst(argtypes[2])
+        # Check that we can determine the element type
+        (isa(a, DataType) && (isa(a.parameters[1], Type) || isa(a.parameters[1], TypeVar))) || return false
+        at = a.parameters[1]
+        # Check that the element type is compatible with the element we're assigning
+        (argtypes[3] ⊑ (isa(at, Type) ? at : at.lb)) || return false
```

Comment:
`lb` might be a TypeVar

Reply:
So, iterate until we hit a type? Can they be circular?

Reply:
can't be circular, I think Jeff added a helper function to unwrap TypeVar